examples/all-clusters-app/all-clusters-common/src/DeviceEnergyManagementDelegateImpl.cpp

```diff
@@ -0,0 +1,250 @@
+    // TODO: implement
+    mEsaState = ESAStateEnum::kOnline;
+    MatterReportingAttributeChangeCallback(mEndpointId, DeviceEnergyManagement::Id, ESAState::Id);
+
+    return status;
+}
+
+Status DeviceEnergyManagementDelegate::ModifyForecastRequest(
+    const uint32_t forecastId, const DataModel::DecodableList<Structs::SlotAdjustmentStruct::DecodableType> & slotAdjustments)
+{
+    Status status = Status::UnsupportedCommand; // Status::Success;
+    // TODO: implement
+    return status;
+}
+
+Status DeviceEnergyManagementDelegate::RequestConstraintBasedForecast(
+    const DataModel::DecodableList<Structs::ConstraintsStruct::DecodableType> & constraints)
+{
+    Status status = Status::UnsupportedCommand; // Status::Success;
+    // TODO: implement
+    return status;
+}
+
+// ------------------------------------------------------------------
+// Get attribute methods
+ESATypeEnum DeviceEnergyManagementDelegate::GetESAType()
+{
+    return mEsaType;
+}
+
+bool DeviceEnergyManagementDelegate::GetESACanGenerate()
+{
+    return mEsaCanGenerate;
+}
+
+ESAStateEnum DeviceEnergyManagementDelegate::GetESAState()
+{
+    return mEsaState;
+}
+
+int64_t DeviceEnergyManagementDelegate::GetAbsMinPower()
+{
+    return mAbsMinPower;
+}
+
+int64_t DeviceEnergyManagementDelegate::GetAbsMaxPower()
+{
+    return mAbsMaxPower;
+}
+
+Attributes::PowerAdjustmentCapability::TypeInfo::Type DeviceEnergyManagementDelegate::GetPowerAdjustmentCapability()
+{
+    return mPowerAdjustmentCapability;
+}
+
+DataModel::Nullable<Structs::ForecastStruct::Type> DeviceEnergyManagementDelegate::GetForecast()
+{
+    return mForecast;
+}
+
+// ------------------------------------------------------------------
+// Set attribute methods
+
+CHIP_ERROR DeviceEnergyManagementDelegate::SetESAType(ESATypeEnum newValue)
+{
+    ESATypeEnum oldValue = mEsaState;
+
+    if (newValue >= ESATypeEnum::kUnknownEnumValue)
+    {
+        return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+    }
+
+    mEsaType = newValue;
+    if (oldValue != newValue)
+    {
+        ChipLogDetail(AppServer, "mEsaType updated to %d", (int) mEsaType);
+        MatterReportingAttributeChangeCallback(mEndpointId, DeviceEnergyManagement::Id, ESAType::Id);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeviceEnergyManagementDelegate::SetESACanGenerate(bool newValue)
+{
+    bool oldValue = mEsaCanGenerate;
+
+    mEsaCanGenerate = newValue;
+    if (oldValue != newValue)
+    {
+        ChipLogDetail(AppServer, "mEsaCanGenerate updated to %d", (int) mEsaCanGenerate);
+        MatterReportingAttributeChangeCallback(mEndpointId, DeviceEnergyManagement::Id, ESACanGenerate::Id);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeviceEnergyManagementDelegate::SetESAState(ESAStateEnum newValue)
+{
+    ESAStateEnum oldValue = mEsaState;
+
+    if (newValue >= ESAStateEnum::kUnknownEnumValue)
+    {
+        return CHIP_IM_GLOBAL_STATUS(ConstraintError);
```

Comment:
If oldValue == newValue
should you raise  ConstraintError
or just quietly ignore it ?